# 에러⚡️

<br/>

### 에러처리를 왜 해야하나요?⚡️

에러를 대처하지 않고 방치하면 프로그램이 강제 종료된다.  
에러가 발생하지 않는 코드를 작성하는 것이 이상적이지만 그것은 불가능하고,  
예외적인 상황은 너무 다양하므로 에러를 전제하고 대응하는 코드를 작성하는 것이 중요하다.

<br/>

---

### 자바스크립트에서 에러를 처리하는 방법에는 뭐가 있을까요?

에러 처리를 구현하는 방법은 크게 두 가지가 있다.

1. querySelector나 Array#find 메서드처럼 예외적인 상황이 발생하면 반환하는 값(null 또는 -1)을 if 문이나 단축 평가 또는 옵셔널 체이닝 연산자를 통해 확인해서 처리하는 방법
2. 에러 처리 코드를 미리 등록해 두고 에러가 발생하면 에러 처리 코드로 점프하도록 하는 방법

- **try... catch... finally 문**  
  2의 방법으로, 일반적으로 이 방법을 에러 처리(error handling)라고 한다.  
  try... catch... finally 문은 다음과 같이 3개의 코드 블록으로 구성되며, finally 문은 생략할 수 있다.

  ```javascript
  try {
    // 실행할 코드(에러가 발생할 가능성이 있는 코드)
  } catch (err) {
    // try 코드 블록에서 에러가 발생하면 이 코드 블록의 코드가 실행된다.
    // err에는 try 코드 블록에서 발생한 Error 객체가 전달된다.
  } finally {
    // 에러 발생과 상관없이 반드시 한 번 실행된다.
  }
  ```

  try 코드 블록이 실행되고 에러가 발생하면 catch 문의 err 변수에 전달되고 catch 코드 블록이 실행된다.

- **Error 객체**  
  Error 생성자 함수는 에러 객체를 생성한다.  
  Error 생성자 함수에는 에러를 상세히 설명하는 에러 메시지를 인수로 전달할 수 있다.

  ```javascript
  const error = new Error('invalid');
  ```

  자바스크립트는 Error 생성자 함수를 포함해 7가지의 에러 객체를 생성할 수 있는 Error 생성자 함수를 제공한다.  
  Error 생성자 함수가 생성한 에러 객체의 프로토타입은 모두 Error.prototype을 상속받는다.

  | 생성자 함수    | 인스턴스                                                                       |
  | :------------- | :----------------------------------------------------------------------------- |
  | Error          | 일반적 에러 객체                                                               |
  | SyntaxError    | 자바스크립트 문법에 맞지 않는 문을 해석할 때 발생하는 에러 객체                |
  | ReferenceError | 참조할 수 없는 식별자를 참조했을 때 발생하는 에러 객체                         |
  | TypeError      | 피연산자 또는 인수의 데이터 타입이 유효하지 않을 때 발생하는 에러 객체         |
  | RangeError     | 숫자값의 허용 범위를 벗어낫을 때 발생하는 에러 객체                            |
  | URIError       | encodeURI 또는 decodeURI 함수에 부적절한 인수를 전달했을 때 발생하는 에러 객체 |
  | EvalError      | eval 함수에서 발생하는 에러 객체                                               |

- **throw 문**
  Error 생성자 함수로 에러 객체를 생성한다고 에러가 발생하는 것은 아니다.  
  에러 객체 생성과 에러 발생은 의미가 다르다.

  ```javascript
  try {
    // 에러 객체를 생성한다고 에러가 발생하는 것은 아니다
    new Error('something wrong');
  } catch (error) {
    console.log(error);
  }
  ```

  에러를 발생시키려면 try 코드 블록에서 throw 문으로 에러 객체를 던져야 한다.

  ```javascript
  throw 표현식;
  ```

  ```javascript
  try {
    // 에러 객체를 던지면 catch 코드 블록이 실행되기 시작한다.
    throw new Error('something wrong');
  } catch (error) {
    console.log(error);
  }
  ```

<br/>

---
